File: docs/lib/catalogues/service_catalogue.js
import { CatalogueConnector } from "../abstractions/catalogue_connector.js";
import { Change } from "../change.js";
import { DefaultChunkSize } from "../constants.js";
import { Page } from "../page.js";
import { Result } from "../result.js";
import { CatalogueConnectorUtils } from "../utils/catalogue_connector_utils.js";

/**
 * @template TItem 
 * @template TFilters
 * @implements {CatalogueConnector<TItem, TFilters>}
 */
class ServiceCatalogue extends CatalogueConnector {

    /** @type {CatalogueConnector<TItem, TFilters>} */
    #localConnector;

    /** @type {CatalogueConnector<TItem, TFilters>} */
    #remoteConnector;

    /** @type {Change[]} */
    #changes;

    /** @type {CatalogueConnector<TItem, TFilters>} */
    get localConnector() { return this.#localConnector; }

    /** @type {CatalogueConnector<TItem, TFilters>} */
    get remoteConnector() { return this.#remoteConnector; }

    /** @type {Change[]} */
    get changes() { return this.#changes; }

    /** @returns {boolean} */
    get available() { return this.#remoteConnector.available || this.#localConnector.available; }

    /**
     * @param {string} id 
     * @param {TItem?} item 
     * @returns {Promise<Result<TItem?>>}
     */
    async onSyncRemoteAsync(id, item) { return await CatalogueConnectorUtils.syncItem(this.#remoteConnector, id, item); }

    /**
     * @param {string} id 
     * @param {TItem?} item 
     * @returns {Promise<Result<TItem?>>}
     */
    async onSyncLocalAsync(id, item) { return await CatalogueConnectorUtils.syncItem(this.#localConnector, id, item); }

    async syncChangesAsync() {
        if (this.#remoteConnector.available)
            try {
                for (let change of [...this.#changes]) {
                    let get_result = await this.#localConnector.getAsync(change.id);
                    let sync_result = await this.onSyncRemoteAsync(change.id, get_result.data);
                    if (sync_result.ok) {
                        await this.onSyncLocalAsync(change.id, sync_result.data);
                        this.#changes.splice(this.#changes.indexOf(change), 1)
                    }
                    else change.errors = sync_result.errors;
                }
            }
            catch (error) { console.log(error); }
    }

    /**
      * @param {TFilters?} filters 
      * @param {number} pageIndex 
      * @param {number} pageSize 
      * @returns {Promise<Result<Page<TItem>?>>}
      */
    async queryAsync(filters = null, pageIndex = 0, pageSize = DefaultChunkSize) {
        if (this.#remoteConnector.available)
            try {
                let result = await this.#remoteConnector.queryAsync(filters, pageIndex, pageSize);
                if (result.data?.items.length > 0) for (let item of result.data?.items) await this.onSyncLocalAsync(item.id, item);
                return result;
            }
            catch (error) { console.log(error); }
        return await this.#localConnector.queryAsync(filters, pageIndex, pageSize);
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<TItem?>>}
     */
    async getAsync(id) {
        if (this.#remoteConnector.available)
            try {
                let result = await this.#remoteConnector.getAsync(id);
                if (result.ok) await this.onSyncLocalAsync(id, result.data);
                return result;
            }
            catch (error) { console.log(error); }
        return await this.#localConnector.getAsync(id);
    }

    /**
     * @param {TItem} item 
     * @returns {Promise<Result<string>>}
     */
    async insertAsync(item) {
        let result = null;
        if (this.#remoteConnector.available)
            try {
                result = await this.#remoteConnector.insertAsync(item);
                if (result.ok) await this.onSyncLocalAsync(id, result.data);
                return result;
            }
            catch (error) { console.log(error); }
        result = await this.#localConnector.insertAsync(item);
        if (result.Ok) this.#changes.push(new Change({ id: item.id }));
        return result;
    }

    /**
     * @param {string} id 
     * @param {TItem} item 
     * @returns {Promise<Result<boolean>>}
     */
    async updateAsync(id, item) {
        let result = null;
        if (this.#remoteConnector.available)
            try {
                result = await this.#remoteConnector.updateAsync(id, item);
                if (result.ok) await this.onSyncLocalAsync(id, result.data);
                return result;
            }
            catch (error) { console.log(error); }
        result = await this.#localConnector.updateAsync(id, item);
        if (result.Ok) this.#changes.push(new Change({ id: id }));
        return result;
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<boolean>>}
     */
    async deleteAsync(id) {
        let result = null;
        if (this.#remoteConnector.available)
            try {
                result = await this.#remoteConnector.deleteAsync(id);
                if (result.ok) await this.onSyncLocalAsync(id, null);
                return result;
            }
            catch (error) { console.log(error); }
        result = await this.#localConnector.deleteAsync(id);
        if (result.Ok) this.#changes.push(new Change({ id: id }));
        return result;
    }

    /**
     * 
     * @param {CatalogueConnector<TItem, TFilters>} localConnector 
     * @param {CatalogueConnector<TItem, TFilters>} remoteConnector 
     * @param {Change[]} changes 
     */
    constructor(localConnector, remoteConnector, changes) {
        super();
        this.#localConnector = localConnector;
        this.#remoteConnector = remoteConnector;
        this.#changes = changes;
    }

}

export {
    ServiceCatalogue
}

File: Core/Catalogues/ServiceCatalogue.cs
using KolibSoft.Catalogue.Core.Abstractions;
using KolibSoft.Catalogue.Core.Utils;
using static KolibSoft.Catalogue.Core.Constants;

namespace KolibSoft.Catalogue.Core.Catalogues;

public class ServiceCatalogue<TItem, TFilters> : ICatalogueConnector<TItem, TFilters>
    where TItem : IItem
{

    public ICatalogueConnector<TItem, TFilters> LocalConnector { get; }
    public ICatalogueConnector<TItem, TFilters> RemoteConnector { get; }
    public virtual bool Available => RemoteConnector.Available || LocalConnector.Available;

    public ICollection<Change> Changes { get; }

    protected virtual Task<Result<TItem?>> OnSyncRemoteAsync(Guid id, TItem? item) => RemoteConnector.SyncItem(id, item);
    protected virtual Task<Result<TItem?>> OnSyncLocalAsync(Guid id, TItem? item) => LocalConnector.SyncItem(id, item);

    public virtual async Task SyncChangesAsync()
    {
        if (RemoteConnector.Available)
            try
            {
                foreach (var change in Changes.ToArray())
                {
                    var get_result = await LocalConnector.GetAsync(change.Id);
                    var sync_result = await OnSyncRemoteAsync(change.Id, get_result.Data);
                    if (sync_result.Ok)
                    {
                        await OnSyncLocalAsync(change.Id, sync_result.Data);
                        Changes.Remove(change);
                    }
                    else change.Errors = sync_result.Errors;
                }
            }
            catch (Exception e) { Console.WriteLine(e.Message); }
    }

    public virtual async Task<Result<Page<TItem>?>> QueryAsync(TFilters? filters = default, int pageIndex = 0, int pageSize = DefaultChunkSize)
    {
        if (RemoteConnector.Available)
            try
            {
                var result = await RemoteConnector.QueryAsync(filters, pageIndex, pageSize);
                if (result.Data?.Items.Any() == true) foreach (var item in result.Data.Items) await OnSyncLocalAsync(item.Id, item);
                return result;
            }
            catch (Exception e) { Console.WriteLine(e.Message); }
        return await LocalConnector.QueryAsync(filters, pageIndex, pageSize);
    }

    public virtual async Task<Result<TItem?>> GetAsync(Guid id)
    {
        if (RemoteConnector.Available)
            try
            {
                var result = await RemoteConnector.GetAsync(id);
                if (result.Ok) await OnSyncLocalAsync(id, result.Data);
                return result;
            }
            catch (Exception e) { Console.WriteLine(e.Message); }
        return await LocalConnector.GetAsync(id);
    }

    public virtual async Task<Result<TItem?>> InsertAsync(TItem item)
    {
        Result<TItem?> result;
        if (RemoteConnector.Available)
            try
            {
                result = await RemoteConnector.InsertAsync(item);
                if (result.Ok) await OnSyncLocalAsync(item.Id, result.Data);
                return result;
            }
            catch (Exception e) { Console.WriteLine(e.Message); }
        result = await LocalConnector.InsertAsync(item);
        if (result.Ok && !Changes.Any(x => x.Id == item.Id)) Changes.Add(new Change(item.Id));
        return result;
    }

    public virtual async Task<Result<TItem?>> UpdateAsync(Guid id, TItem item)
    {
        Result<TItem?> result;
        if (RemoteConnector.Available)
            try
            {
                result = await RemoteConnector.UpdateAsync(id, item);
                if (result.Ok) await OnSyncLocalAsync(id, result.Data);
                return result;
            }
            catch (Exception e) { Console.WriteLine(e.Message); }
        result = await LocalConnector.UpdateAsync(id, item);
        if (result.Ok && !Changes.Any(x => x.Id == id)) Changes.Add(new Change(id));
        return result;
    }

    public virtual async Task<Result<TItem?>> DeleteAsync(Guid id)
    {
        Result<TItem?> result;
        if (RemoteConnector.Available)
            try
            {
                result = await RemoteConnector.DeleteAsync(id);
                if (result.Ok) await OnSyncLocalAsync(id, default);
                return result;
            }
            catch (Exception e) { Console.WriteLine(e.Message); }
        result = await LocalConnector.DeleteAsync(id);
        if (result.Ok && !Changes.Any(x => x.Id == id)) Changes.Add(new Change(id));
        return result;
    }

    public ServiceCatalogue(ICatalogueConnector<TItem, TFilters> localConnector, ICatalogueConnector<TItem, TFilters> remoteConnector, ICollection<Change> changes)
    {
        LocalConnector = localConnector;
        RemoteConnector = remoteConnector;
        Changes = changes;
    }

}